### ISSUE TITLE
Segmentation fault on Migrate() on s390x

### ISSUE THREAD

**Issue by GuillaumeDumo** [dotnet/efcore#32209]
A "Segmentation fault" occurs when trying to run a Migrate() on a s390x.

The last trace I get is this.

> dbug: Microsoft.EntityFrameworkCore.Database.Command[20100]
      Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='30']
      PRAGMA journal_mode = 'wal';

### Code

```C#
using Microsoft.EntityFrameworkCore;

namespace ConsoleApp1
{
    internal sealed class Program
    {
        static void Main(string[] args)
        {
            WebApplicationBuilder builder = WebApplication.CreateBuilder(args);
            builder.Services.AddDbContext<MyDbContext>(optionsAction =>
            {
                optionsAction.UseSqlite("Data Source=/App_Data/Mimic.db");
            });

            WebApplication app = builder.Build();

            using (IServiceScope scope = app.Services.CreateScope())
            {
                if (Directory.Exists("App_Data") is false)
                {
                    Directory.CreateDirectory("App_Data");
                }
                var db = scope.ServiceProvider.GetRequiredService<MyDbContext>();
                db.Database.Migrate();
            }
        }
    }

    public class Test
    {
        public int Id { get; set; }
    }

    public class MyDbContext : DbContext
    {
        public MyDbContext(DbContextOptions<MyDbContext> dbContextOptions) : base(dbContextOptions)
        {
        }
        public DbSet<Test> Tests => Set<Test>();
    }
}
```

### Fulll run
```
dbug: Microsoft.EntityFrameworkCore.Infrastructure[10401]
      An 'IServiceProvider' was created for internal use by Entity Framework.
dbug: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 7.0.13 initialized 'MyDbContext' using provider 'Microsoft.EntityFrameworkCore.Sqlite:7.0.13' with options: None
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20005]
      Creating DbConnection.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20006]
      Created DbConnection. (16ms).
dbug: Microsoft.EntityFrameworkCore.Migrations[20400]
      Migrating using database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20005]
      Creating DbConnection.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20006]
      Created DbConnection. (1ms).
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20000]
      Opening connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20004]
      An error occurred using the connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20007]
      Disposing connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20008]
      Disposed connection to database 'main' on server '/App_Data/Mimic.db' (0ms).
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20005]
      Creating DbConnection.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20006]
      Created DbConnection. (0ms).
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20000]
      Opening connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20004]
      An error occurred using the connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20007]
      Disposing connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20008]
      Disposed connection to database 'main' on server '/App_Data/Mimic.db' (0ms).
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20000]
      Opening connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Connection[20001]
      Opened connection to database 'main' on server '/App_Data/Mimic.db'.
dbug: Microsoft.EntityFrameworkCore.Database.Command[20103]
      Creating DbCommand for 'ExecuteNonQuery'.
dbug: Microsoft.EntityFrameworkCore.Database.Command[20104]
      Created DbCommand for 'ExecuteNonQuery' (5ms).
dbug: Microsoft.EntityFrameworkCore.Database.Command[20106]
      Initialized DbCommand for 'ExecuteNonQuery' (9ms).
dbug: Microsoft.EntityFrameworkCore.Database.Command[20100]
      Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='30']
      PRAGMA journal_mode = 'wal';

=================================================================
        Native Crash Reporting
=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

=================================================================
        Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x10042):0x10032  Segmentation fault (core dumped)
```

Microsoft.EntityFrameworkCore.Sqlite: 7.0.13
Microsoft.EntityFrameworkCore.Design: 7.0.13
.NET 6.0
Operating system:
alpine:3.17.5 on  s390x

I also tried using SqliteConnection. I found out that if I try to create table i get the same Segmentation fault when executing the query, but if I do a select without "from" it works fine. So the issue may be on the writing to the .db file.

```C#
 using (var connection = new SqliteConnection(builder.Configuration.GetConnectionString("sqlite")))
            {
                if (Directory.Exists("App_Data") is false)
                {
                    Directory.CreateDirectory("App_Data");
                }

                Console.WriteLine("connection.Open s");
                connection.Open();
                Console.WriteLine("connection.Open e");

                var command = connection.CreateCommand();
                command.CommandText =
                @"
                Create table if not exists Tests
                (
                    Id int
                )
            ";
                Console.WriteLine("ExecuteNonQuery s");
                command.ExecuteNonQuery();
                Console.WriteLine("ExecuteNonQuery e");

            }
```

```C#
 using (var connection = new SqliteConnection(builder.Configuration.GetConnectionString("sqlite")))
            {           
                connection.Open();

                var command = connection.CreateCommand();
                command.CommandText =
                @"SELECT $id";
                command.Parameters.AddWithValue("$id", Random.Shared.Next());

                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        var id = reader.GetInt32(0);

                        Console.WriteLine($"Hello, {name}!");
                    }
                }
            }
```

**Comment by roji** [comment]
@bricelam s390x is notably big endian; I wonder if the right SQLite native binaries are getting used etc.

**Comment by uweigand** [comment]
The SQLite binaries are native machine code, they wouldn't run at all if they weren't s390x ...  But of course there can still be big-endian bugs in them somewhere.

@giritrivedi @saitama951 can you try to recreate this on .NET 6 on RHEL ?

@GuillaumeDumo do you know whether this worked with the 6.0 versions of the efcore packages?

**Comment by GuillaumeDumo** [comment]
@uweigand I was previously on 6.0.23, I updated to the latest in my debug process to see if it was something already fixed.

**Comment by saitama951** [comment]
Hi @GuillaumeDumo ,
I tried to repro the steps as mentioned, this seems to work fine on the rhel 9  with net6.0
```
using System;
using System.IO;
using Microsoft.Data.Sqlite;

namespace HelloWorldSample
{
    class Program
    {
        static void Main()
        {
            using (var connection = new SqliteConnection("Data Source=hello.db"))
            {
                connection.Open();

                var command = connection.CreateCommand();
                command.CommandText =
                @"
                        Create table if not exists Tests
                (
                    Id int
                )
                ";
                command.ExecuteNonQuery();
                Console.WriteLine("command successfull");
           }
        }
     }
}
```
output

```
[root@saitama repro1]# dotnet run
command successfull
[root@saitama repro1]#
```
SDK details
```
.NET SDK (reflecting any global.json):
 Version:   6.0.200
 Commit:    4c30de7899

Runtime Environment:
 OS Name:     rhel
 OS Version:  9
 OS Platform: Linux
 RID:         linux-s390x
 Base Path:   /root/sanjam/issue/dotnet-s390x/sdk/6.0.200/

Host (useful for support):
  Version: 6.0.2
  Commit:  N/A

.NET SDKs installed:
  6.0.200 [/root/sanjam/issue/dotnet-s390x/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.2 [/root/sanjam/issue/dotnet-s390x/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.2 [/root/sanjam/issue/dotnet-s390x/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
  ```

**Comment by uweigand** [comment]
If it works on RHEL but crashes on Alpine, maybe there's an OS-specific (or musl libc-specific?) problem somewhere.

@ayakael, does that ring any bells?

**Comment by bricelam** [comment]
We've seen lots of segfaults on alpine/musl over the years with SQLite. They're usually caused by a missing or misaligned dependency. Does it work if you install the sqlite-dev package?

**Comment by ayakael** [comment]
> Referenc

Havn't seen this one in the wild yet, unfortunately.

**Comment by GuillaumeDumo** [comment]
Adding  sqlite-dev and/or sqlite-libs did not fix the issue but I change from alpine to fedora and it fix the issue.

I'm still not sure what the source of the issue was but looking at the date of the dotnet package on alpine, dotnet-sdk is on 6.0.115 and runtime is 6.0.15 (https://dl-cdn.alpinelinux.org/alpine/v3.17/community/s390x/), it is possible that there is some other packages also out of date.

**Comment by uweigand** [comment]
The crash happens in the libe_sqlite3.so native library provided by the SQLitePCLRaw.lib.e_sqlite3 nuget package.  I don't fully understand the crash, but it happens as that library is calling some libc routines.  As the package is (cross-)built against glibc, there may be some musl incompatibility here.

I note that SQLitePCLRaw.lib.e_sqlite3 provides *different* versions of that libe_sqlite3.so today for linux-x64 vs. linux-musl-x64 and similarly for Arm - but on s390x we do not have a special musl build.   So I'm now trying to see if we can get that added same as for the other Linux targets, and see whether this fixes the crash.

**Comment by uweigand** [comment]
>So I'm now trying to see if we can get that added same as for the other Linux targets, and see whether this fixes the crash.

Indeed a musl-specific build of `libe_sqlite3.so` fixes the crash in local testing.  I've opened an upstream issue: https://github.com/ericsink/SQLitePCL.raw/issues/571

**Comment by uweigand** [comment]
`SQLitePCL.raw` support for `linux-musl-s390x` has now been merged upstream and should be available in the next release (v2.1.7).

**Comment by uweigand** [comment]
`SQLitePCLRaw.core` version 2.1.7 including support for `linux-musl-s390x` has now been released:
https://www.nuget.org/packages/SQLitePCLRaw.core
